help/en/html/tools/logixng/Tutorial/chapter10.shtml

```diff
@@ -0,0 +1,220 @@
+<!DOCTYPE html>
+<html lang="en">
+
+<head>
+  <title>LogixNG Tutorial - Chapter 10</title>
+  <meta name="author" content="Daniel Bergqvist">
+  <meta name="author" content="Dave Sand">
+  <meta name="keywords" content="jmri LogixNG tutorial modules">
+
+  <!--#include virtual="/Style.shtml" -->
+
+</head>
+
+<body>
+  <!--#include virtual="/Header.shtml" -->
+
+  <div id="mBody">
+
+    <div id="mainContent" class="no-sidebar">
+      <h1>LogixNG Tutorial - Chapter 10</h1>
+
+      <h2>Modules</h2>
+
+      <p>A module is a special ConditionalNG which is invoked from regular ConditionalNGs.
+      In programing terms it is a sub-routine, method, def, etc.  It can use zero or more
+      inputs and return zero or more outputs.  It does not respond to expression changes.</p>
+
+      <p>A common use would be to return the state of a set of expressions that can be
+      called by various ConditionalNGs rather than listing each expressions in each
+      ConditionalNG.  Another exmaple is setting a common group of actions.</p>
+
+      <p>When combined with <a href="chapter11.shtml">tables</a>, a module could be passed
+      the name of a table row.  It would then perform the actions based on the table
+      row data.</p>
+
+      <p>The <a href="/help/en/package/jmri/jmrit/beantable/LogixNGModuleTable.shtml">
+      Module Table</a> provides the details on creating a module.</p>
+
+      <p>This is a digital expression socket.  It can have inputs, but it is limited to
+      returning <strong>true</strong> or <strong>false</strong> based on the
+      <strong><em>current</em></strong> state of the expression list.</p>
+
+      <div style="margin-left: 2em;">
+      <a href="images/chapter10/digital_expression_root.png">
+      <img src="images/chapter10/digital_expression_root.png"
+      alt="Chapter 10 digital expression root" width="500" height="108">
+      </a>
+      </div>
+
+      <p>This is a digital action socket.  This can be used to create an action only module or
+      complex modules withe any combination of expressions and actions, such as
+      <strong>if then else</strong> structures.  These can have inputs and outputs.</p>
```

Comment:
When I first read this sentense, I got the feeling that the sentense `These can have inputs and outputs` is related to a digital action socket module. I suggest that you put this sentence in it's own paragraph, since digital expressions can have inputs and outputs as well.

Reply:
Can a digital expression return a value as well as true/false?

Reply:
Yes. I might not have tested that, but in this regard there should be no difference between digital action modules and digital expression modules.

Reply:
Think of Java. You could have a method:
`boolean isValid(Object a, List<String> errors);`

which checks if `a` is valid in some sense, and if not, returns false and adds the error messages in the list `errors`.